[FLINK-2399] Version checks for Job Manager and Task Manager

sachingoel0101 opened the pull request:
[Reopening #944. The travis build failed somehow. ]
Both Job Manager and Task Manager now communicate their respective version IDs to each other as part of registration messages.
Version ID is accessed by the information Maven adds in the manifest.

StephanEwen commented:
I think we need to decide first what we actually mean by version and what is supposed to match what.

What is actually the value `getClass().getPackage().getImplementationVersion()`?

sachingoel0101 commented:
I had decided to work with my own understanding of what version means since nobody replied to the JIRA comment.
getClass.getPackage.getImplementationVersion would return, 0.10 SNAPSHOT for the current build. In case this string doesn't match for the TM and JM, it would fail the registration process.

StephanEwen commented:
Using `getClass().getPackage().getImplementationVersion()` would be a decent first approach then, I guess. The critical part seems to be the Client-to-JobManager communication.

How about the following as a first step: The client sends its version together with the `SubmitJob` message (just add a field there). The JobManager would check the version and respond with a failure, if it does not match. You can probably make the JobManager part very simple, no need to add extra constructor parameters, etc.

That way, the change would be minimally invasive, and we could see how well it addresses the issues, and whether we should extend this to other messages as well.

sachingoel0101 commented:
1. Added version checks between JobClient and JobManager. 
2. Versions are accessed from the Configuration class now, since flink-core gets built first and version is readily available, leading to exact verification.
3. Added a dummy version string to allow passing of tests in IDE.

sachingoel0101 commented:
I've already added a version check between `JobClient` and `JobManager`. Will there be any further review of this?

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -121,11 +121,14 @@ class JobManager(
 
   override val leaderSessionID = Some(UUID.randomUUID())
 
+  private var jobManagerVersionID: String = _
```
Better to use `Option` instead of `null`.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -166,7 +169,8 @@ class JobManager(
       hardwareInformation,
-      numberOfSlots) =>
+      numberOfSlots,
+      taskManagerVersionID) =>
```
Why do you check the version when a `TaskManager` registers at the `JobManager`. I thought the critical part would be the `JobClient` and `JobManager` communication.

sachingoel0101 replied:
This was to ensure that the binaries on all machines are the same.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -184,26 +188,32 @@ class JobManager(
+            // to be notified when the taskManager is no longer reachable
+            context.watch(taskManager)
+          } else{
+            throw new Exception("Version mismatch error between Job Manager and Task Manager")
```
This will kill the `JobManager`. This is not a good solution. Best you revert the verification between the `TaskManager` and the `JobManager`.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -524,13 +534,24 @@ class JobManager(
+      sender ! decorateMessage(
+        Failure(
+          new JobSubmissionException(jobGraph.getJobID, "Client version mismatches that of Job " +
+            "Manager.")
```
What versions do the client and the job manager have?

sachingoel0101 replied:
The version is drawn from the Jar file's package version. Maven automatically adds this info.

tillrohrmann replied:
Yeah I know, but it might be interesting for the user who sees the `JobSubmissionException` to know the exact version.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -547,7 +552,7 @@ class TaskManager(
         // successful registration. associate with the JobManager
         // we disambiguate duplicate or erroneous messages, to simplify debugging
-        case AcknowledgeRegistration(_, leaderSessionID, jobManager, id, blobPort) =>
+        case AcknowledgeRegistration(_, leaderSessionID, jobManager, id, blobPort, jobManagerID) =>
```
This does not work because a `TaskManager's` registration can also be acknowledged by an `AlreadyRegistered` message.

sachingoel0101 replied:
But the initial registration must have happened with `AcknowledgeRegistration`, right? Correct me if I'm wrong.

tillrohrmann replied:
The initial AcknowledgeRegistration message can get lost and then the JM will send a `AlreadyRegistered` message.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobSubmitTest.java:
```diff
@@ -170,4 +170,43 @@ public void initializeOnMaster(ClassLoader loader) throws Exception {
+		}
+		catch (Exception e) {
+			e.printStackTrace();
+			fail(e.getMessage());
```
Why do you print the stack trace and let the test fail with an `AssertionError` instead of letting the test fail with the true reason, e.g. `e`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobSubmitTest.java:
```diff
@@ -170,4 +170,43 @@ public void initializeOnMaster(ClassLoader loader) throws Exception {
+			catch (JobExecutionException e) {
+				// that is what we expect
+				// test that the exception nesting is not too deep
+				assertTrue(e.getCause() == null);
```
Better to test for the proper exception message.

sachingoel0101 replied:
Yeah. I'll push a fix.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobSubmitTest.java:
```diff
@@ -170,4 +170,43 @@ public void initializeOnMaster(ClassLoader loader) throws Exception {
+
+				@Override
+				public void initializeOnMaster(ClassLoader loader) throws Exception {
+					throw new RuntimeException("test exception");
```
Why using a faulty `JobVertex` here. This might also be the reason for the "expected" `JobExecutionException` then. And you don't check properly whether that's not the cause for it.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/configuration/Configuration.java:
```diff
@@ -54,8 +54,12 @@
+
+	/** Stores the program version */
+	public final String FLINK_VERSION = getClass().getPackage().getImplementationVersion() ==
+		null ? "FLINK_TEST_VERSION" : getClass().getPackage().getImplementationVersion();
```
What's the value of `FLINK_VERSION` if you execute a Flink job on a standalone cluster from your IDE using the `RemoteExecutionEnvironment`? Will it work?

tillrohrmann replied:
What happens with minor versions which are compatible, e.g. 0.9.0 and 0.9.1?

sachingoel0101 replied:
Ah yes. I had not considered this. In this case, the client version would be null.

As for ensuring that compatible versions still work, you're right. This is why I'm not in favor of using the current version string. A better approach would be to have specific release versions in the configuration file, from where everything starts.`

sachingoel0101 replied:
In such a case, the remote execution problem gets solved, since the version string would be properly set.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClient.java:
```diff
@@ -205,7 +205,7 @@ public static void submitJobDetached(
 
 		Future<Object> future = jobManagerGateway.ask(
-				new JobManagerMessages.SubmitJob(jobGraph, false),
+				new JobManagerMessages.SubmitJob(jobGraph, false, jobGraph.getClientVersion()),
```
Why don't you simply use the `getImplementationVersion` of the `JobClient` but the version of the `Configuration` which is part of the `jobGraph` instance?

tillrohrmann commented:
I had some comments concerning the implementation. I think as a first step we should not check whether the JM and TM have the same versions because in nearly all cases both classes will be started from the same binaries. If we should decide at a later point that we want to check that all used actors are compatible, then I would vote for using the decoration mechanism to transparently add a version ID instead of adding it manually to a subset of messages. The latter approach is IMO error-prone.

Checking whether the client and the JM have compatible versions should be sufficient. I'm not quite sure whether `getClass().getPackage().getImplementationVersion()` is the right way to go here. Using this function won't allow compatible versions which only differ in a minor version number to work together. IMHO, this should be allowed.

sachingoel0101 commented:
Hi @tillrohrmann , thanks for the review. You're right. `getClass...` is not the right way to go. I had decided to just use this for the moment. We can certainly have a better version numbering for this purpose, to ensure the actors are compatible.
I just am not sure if having a version string, which needs to be managed directly in the source code is a good idea. For example, it would make sense to extract the first digit, 0.9, 0.10 etc. since the minor bug fix releases are otherwise compatible anyway.

sachingoel0101 commented:
@tillrohrmann , I have modified the way version information is passed to the Job Manager. It is now through the Configuration object passed.
Further, rolled back on any verification for Job Manager and Task Manager versions.
There is a possibility to handle compatibility between different versions by specifying a minimum supported version information. Note that this will easily be adaptable to Task manager version verification, if needed at some point. 
Let me know if this is a better solution.

sachingoel0101 commented:
Rebased to the current master.

fhueske commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobSubmitTest.java:
```diff
@@ -192,4 +193,78 @@ public void initializeOnMaster(ClassLoader loader) throws Exception {
+	/**
+	 * Verifies failure when the client version is lower than supported.
+	 */
+	@Test
```
You can implement the test with `@Test(expected=JobSubmissionException.class)` which checks that the correct exception is produced.

fhueske commented on flink-core/src/main/java/org/apache/flink/util/VersionUtils.java:
I doubt that it is OK to copy code from Stackoverflow. SO is under Creative Commons license. If I understood CC correctly, it requires that our code must be under CC as well if we want to use it. See http://creativecommons.org/licenses/by-sa/2.5/

sachingoel0101 commented on flink-core/src/main/java/org/apache/flink/util/VersionUtils.java:
Ah. Sorry. Will fix this.

sachingoel0101 commented:
Un-assigning myself from the issue for now.
